1 What is the difference between a function and method in a python.
--> A function is a block of reusable code that performs a specific task.It is defined using def and can be called independently, not tied to any object.
--> A method is also a function but it belongs to an object(or class).
You call it using dot notation on a specific object or instance.

In [3]:
def greet(name):
    return f"Hello,{name}!"

print(greet("soniya"))    

Hello,soniya!


In [4]:
name = "soniya"
print(name.upper()) #Here .upper() is a method that belongs to the string class(string object)

SONIYA


2 Explain the concepts of function arguments and parameters in python
--> Parameters are the placeholders or variables listed inside the parentheses when defining a function. They define what kind of data the function can accept.
--> Arguments are the actual values you pass toa function when you call it.They are given to the paraneters of the function.

In [5]:
def greet(name):  #here name is parameter-it's part of the function definition
    print("Hello", name) 

greet("soniya")  #"Soniya" is an argument-it's the actual data passed into the parameter name.
    

3 What are the different ways to define and call a function

Function with no parameters: A function that doesn't take any input

In [1]:
def greet():
    print("Hello, welcome to python!")

#calling    
greet()    

Hello, welcome to python!


Function with parameters: A function that accepts inputs(parameter).

In [3]:
def add(a, b):
    print("Sum:", a+b)

#calling
add(5,3)

Sum: 8


4 What is the purpose of return statement in a python statement
--> The purpose of the return statement in python is to send a value back from a function to the part of the program where the function was called.

In [4]:
def add(a, b):
    result = a+b
    return result

sum_result = add(5,3)
print("Sum is: ", sum_result)
    

Sum is:  8


5 What are iterators in python and how do they differ from iterable?

--> An iterable is any object that can be looped over using a for loop. It means you can go through its elements one by one. Whereas an iterator is an object that actually returns items one at a time when you call next() on it.

In [5]:
my_list = [10, 20, 30, 40, 50]
for item in my_list:
    print(item)               #Here my_list is iterable because we can loop through it

10
20
30
40
50


In [6]:
my_list = [10, 20, 30, 40]
my_iterator = iter(my_list)

print(next(my_iterator))
print(next(my_iterator))
print(next(my_iterator))

10
20
30


6 Explain the concept of generators in python and how they are defined

---> A generator in Python is a special type of function that allows you to generate values one at a time instead of storing them all in memory.
---> Generators are useful when you have large amount of data or when you don't need all results at once.
---> You define a generator just like a normal function, but instead of return, you use the yield keyword.


In [7]:
def squares(n):
    for i in range(1, n+1):
        yield i*i
        
s = squares(5)

for num in s:
    print(num)

1
4
9
16
25


7 What are the advantages of using a generator function over a regular function

--> A regular function returns all values at once which takes more memory.
--> A generator produces one value at a time-it does not store the entire sequence in memory.

-->Generators generate values only when needed. Regular functions calculate everything immediately and return all results. This makes generators faster and efficient for large or infinite sequences.

8 What is lambda function in python and when is it typically used?
--> A lambda function is a small aninymous function that is defined using the keyword lambda. It is used when you need a short, simple function for a short period-usually just one line.

lambda arguments: expression
lambda is a keyword, arguments are input values , expression is a single line of code that is evaluated and automatically returned.

In [8]:
add = lambda a, b: a+b
print(add(5,3))

8


Lambda functions are usually used when you want a quick function without defining it using def, especially inside other functions.

9 Explain the purpose and usage of map() function in python

The map() function is used to apply a function to every item of an iterable and return a map object. It allows you to process all elements without using a loop manually.

In [9]:
def square(x):
    return x*x

numbers = [1, 2, 3, 4, 5]
squared_numbers = map(square, numbers)
print(list(squared_numbers))

[1, 4, 9, 16, 25]


10 What is the difference between map(), reduce(), filter() function in python

map() function : Apply a function to every item in an iterable and return a new iterable of results.
filter() function: select items from an iterable that meet a condition.
reduce() function: combine all elements of an iterable into a single value using a binary function.

In [13]:
#map()--> map(function, iterable)
numbers = [1, 2, 3, 4, 5]
doubled = map(lambda x: x*2, numbers)
print(list(doubled))

[2, 4, 6, 8, 10]


In [11]:
#filter()---> filter(function, iterable)
numbers = [1, 2, 3, 4, 5, 6]
even_numbers = filter(lambda x: x % 2 == 0, numbers)
print(list(even_numbers))

[2, 4, 6]


In [12]:
#reduce()---> from functools import reduce
#reduce(function, iterable)

from functools import reduce
numbers = [1, 2, 3, 4]
summation = reduce(lambda x, y: x+y, numbers)
print(summation)


10


11 Using pen and paper write the internal mechanism for sum operation using reduce function on this given list:[47,11,42,13]. Attach paper image for this answer in doc or colab notebook.

In [14]:
#Example:
from functools import reduce
lst = [47, 11, 42, 13]
result = reduce(lambda x,y: x+y, lst)
print(result)

113


Initial list: [47, 11, 42, 13]
step1 : 47 + 11 = 58
step2: 58 + 42 = 100
step3: 100 + 13 = 113

practical questions

1. Write a Python function that takes a list of numbers as input and returns the sum of all even numbers in
the list.

In [1]:
def sum_of_even_numbers(numbers):
    total = 0
    for num in numbers:
        if num % 2 == 0:
            total += num
    return total
l1 = [1, 2, 3, 4, 5, 6]
print(sum_of_even_numbers(l1))

12


2. Create a Python function that accepts a string and returns the reverse of that string

In [12]:
def reverse_string(text):
    return text[::-1]

text1 = "SONIYA"
reverse_string(text1)
    

'AYINOS'

3. Implement a Python function that takes a list of integers and returns a new list containing the squares of
each number.

In [13]:
def squares_of_numbers(numbers):
    squared_list = []
    for num in numbers:
        squared_list.append(num**2)
    return squared_list
    
list = [1, 2, 3, 4, 5, 6, 7]
print("squares:", squares_of_numbers(list))
    
        

squares: [1, 4, 9, 16, 25, 36, 49]


In [ ]:
4. Write a Python function that checks if a given number is prime or not from 1 to 200.

In [4]:
def is_prime(n):
    return n>1 and n<=100 and all(n% i !=0 for i in range(2, n))
print(is_prime(37))

True


In [ ]:
5. Create an iterator class in Python that generates the Fibonacci sequence up to a specified number of
terms.

In [ ]:
6. Write a generator function in Python that yields the powers of 2 up to a given exponent.

In [8]:
def powers_of_two(n):
    for i in range(n+1):
        yield 2**i

for value in powers_of_two(5):
    print(value)

1
2
4
8
16
32


In [ ]:
7. Implement a generator function that reads a file line by line and yields each line as a string.

In [ ]:
8. Use a lambda function in Python to sort a list of tuples based on the second element of each tuple

In [12]:
data = [(1,5), (3,2), (4,8), (2,1)]
sorted_data = sorted(data, key=lambda x: x[1])
print("sorted list:", sorted_data)

sorted list: [(2, 1), (3, 2), (1, 5), (4, 8)]


In [ ]:
9. Write a Python program that uses `map()` to convert a list of temperatures from Celsius to Fahrenheit.

In [13]:
celsius = [0, 10, 20, 30, 40]
fahrenheit = list(map(lambda c: (c* 9/5) + 32, celsius))
print("Temperatures in Fahrenheit:", fahrenheit)

Temperatures in Fahrenheit: [32.0, 50.0, 68.0, 86.0, 104.0]


In [ ]:
10. Create a Python program that uses `filter()` to remove all the vowels from a given string.

In [16]:
def remove_vowels(text):
    vowels = "aeiouAEIOU"
    return ''.join(filter(lambda ch: ch not in vowels, text))

str1 = "Annapurna"
result = remove_vowels(str1)
print("string without vowels:", result)
    

string without vowels: nnprn


In [ ]:
11) Imagine an accounting routine used in a book shop. It works on a list with sublists, which look like this:



Write a Python program, which returns a list with 2-tuples. Each tuple consists of the order number and the
product of the price per item and the quantity. The product should be increased by 10,- € if the value of the
order is smaller than 100,00 €.

Write a Python program using lambda and map.